HiPE: Support for literal tag, tests and bugfixes

margnus1 opened the pull request:
This PR adds support for tagged literals to HiPE. Tagged literals is one of the mechanisms the Erlang VM can use to identify what data is literal, and what is dynamically allocated. Adding this support will allow HiPE to be used on platforms where this mechanism is used.

I expect the performace to be worse when using literal tagging as the generated code is uglier. The bad code does however give the optimisers some room to stretch their legs, for once. As a result of this, two previously unknown bugs in the Lazy Code Motion optimiser were exposed and fixed.

To be able to reliably reproduce these bugs (as they rely on the garbage collector to hit at the wrong time) the test for these bugs is in the form of a validation pass that is run on the intermediary code as part of the HiPE test suite. The validation pass detects an entire class of bugs, and also serves as a test for the bug fixed by #1607.

sverker commented:
The two bugs in hipe_rtl_lcm; are those real potential problems that deserve fix in maint for 20.2, or is it not worth it.

About `is_divisible`, I think it _has_ the same kind of problem as `bs_validate_unicode` did **in the happy-path**. `bs_validate_unicode` did not _cause_ a GC in happy-path, but it could trigger a pending GC. The reason for this is we changed the condition for GC in OTP-19, but the hipe bif wrappers still use the old condition.
Old condition was to only look at Process.mbuf pointer.
New condition is more complex (https://github.com/erlang/otp/blob/maint/erts/emulator/beam/erl_gc.h#L132), but a heap fragment (mbuf) is not enough.

margnus1 commented:
> The two bugs in hipe_rtl_lcm; are those real potential problems that deserve fix in maint for 20.2, or is it not worth it.

They could certainly affect maint too, but I'm not aware of HiPE generating any code affected by this misoptimisation before this change. There's no reason that it can't happen, though. This code has been around since R10B-2, as far as I can tell, so it would probably have been found if any of the old code triggered the bug. I suppose we have changed code generation recently, but all of that came with tests, so if it can happen on maint, I suspect it'd take some very particular input code.

> we changed the condition for GC in OTP-19, but the hipe bif wrappers still use the old condition

I see, that's why #1607 was so surprising to me. So we should update the bif wrappers at some point to reflect the new logic, I take it.

I'm not changing `is_divisible` due to the change in GC condition (I wasn't even aware), but rather because I felt relying on the happy path only not causing a GC is brittle and smelly. Also, encoding that logic in `hipe_rtl_verify_gcsafe` is inconvenient as some information needed for that is not preserved by the liveness analysis.

sverker commented:
I'm thinking about what to write in release note (if at all).
Did you test some OS that I can mention or did you just force it on in the code?

margnus1 commented:
@sverker Primarily I just forced the literal tag on.

I wrote this code to support a Windows port, and did test it on Windows, but that branch is still plagued by a final tricky bug (unrelated to literal tagging) I haven't found the time to debug, so I don't have an ETA when that feature will land.
